### ISSUE TITLE
The only curve for TLS' ECDH is secp256r1, typically preventing PFS with Prosody

### ISSUE THREAD

**Issue by adrien-n** [siacs/Conversations#2395]
Hi,

Running latest OxygenOS for Oneplus 3T (Android 7), latest Conversations on F-Droid (1.17.0) only offers secp256r1 during the TLS handshake. Prosody uses secp384r1 by default and this makes Conversations typically not get a TLS connection with forward secrecy. Note that ejabberd uses secp256r1 by default I /think/.

Conversations should offer several curves during the TLS handshake in order to accomodate existing servers and allow future migration between curves. I don't know what decides which curves to offer during the client hello.

This can be checked with Wireshark: its TLS dissector will give the curves that are listed in the client and server hello messages.

NB: I don't know if Prosody/Lua-sec(/TLS) can handle several curves on the server side but some things (curve names and doc at least) should be improved on its side too.

**Comment by iNPUTmice** [comment]
The TLS stack is provided by Android. Conversations just removes some very weak ciphers. If Conversations doesn't offer a certain cipher it means your Android distribution doesn't have it.

**Comment by mimi89999** [comment]
@adrien-n You can enable DHE in Prosody. That's one solution.

**Comment by adrien-n** [comment]
A few more infos: I've been told that there was a bug in Android 7.0 and only 7.0 where only P-256 is available. Such an issue had seemed too unrealistic to me to even consider at first but now that I'm on 7.1, I can confirm this was the root cause. I guess only P-256 was tested because that's what is being used for Web browsers.

I really really wanted to avoid DHE and use ECDHE only and it seemed it could work until I expanded my server's federation to an ejabberd server which only uses P-256 while I now use P-384 on mine. Clients and servers for me or around me, that I could improve but not ones further away from me so I had to enable DHE again.

Whenever I get time (hopefully in a few weeks at most), I'll try to add details about the current server connection in Conversations, unless I'm told before that's something that is not welcomed. The motivation is that I had to resort to using Wireshark in order to understand what was going on with the ECDH curves.
